cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -62,6 +62,57 @@ const FunctionDoc list_value_length_doc{
      "Null values emit a null in the output."),
     {"lists"}};
 
+template <typename Type, typename IndexType>
+Status ListElement(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+  using ListArrayType = typename TypeTraits<Type>::ArrayType;
+  using IndexScalarType = typename TypeTraits<IndexType>::ScalarType;
+  ListArrayType list_array(batch[0].array());
+  const auto& index_scalar = batch[1].scalar_as<IndexScalarType>();
+  auto index = index_scalar.value;
+  if (ARROW_PREDICT_FALSE(index < 0)) {
+    return Status::Invalid("Index ", index,
+                           " is out of bounds: should be greater than or equal to 0");
+  }
+  std::unique_ptr<ArrayBuilder> builder;
+  RETURN_NOT_OK(MakeBuilder(ctx->memory_pool(), list_array.value_type(), &builder));
+  for (int i = 0; i < list_array.length(); ++i) {
+    if (list_array.IsNull(i)) {
+      RETURN_NOT_OK(builder->AppendNull());
+      continue;
+    }
+    std::shared_ptr<arrow::Array> value_array = list_array.value_slice(i);
+    auto len = value_array->length();
+    if (ARROW_PREDICT_FALSE(index > len)) {
+      return Status::Invalid("Index ", index, " is out of bounds: should be in [0, ", len,
+                             "]");
+    }
+    ARROW_ASSIGN_OR_RAISE(auto scalar, (value_array->GetScalar(index)));
+    RETURN_NOT_OK(builder->AppendScalar(*scalar));
```

Comment:
You may actually prefer AppendArraySlice which lets you skip allocating a scalar entirely: https://github.com/apache/arrow/blob/012248a7aca9373a39d5ac8ce9e496b8df0f10e6/cpp/src/arrow/array/builder_base.h#L126-L132

Reply:
Can `GetView(i)` be used instead of `GetValue(i)`? Do we want a copy in the output array?

Reply:
I'm not sure what you mean. You have to copy to the output array. GetView may save an allocation vs GetValue, but AppendArraySlice skips those concerns entirely.

Reply:
Oh, I see, thanks for the knowledge bit on `AppendArraySlice`.

Reply:
GetView would certainly be more efficient, but you would have to template the implementation on every possible type, so you're trading off performance for amount of generated code.

Reply:
Thanks!